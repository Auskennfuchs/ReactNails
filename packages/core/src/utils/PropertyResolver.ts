import { css } from 'styled-components'

import { Color } from '../Color'
import {
    BreakPoint, CssFunction, MediaQuery, MediaQueryBreakPoint,
} from '../Layout/MediaQuery'
import { BaseTheme, SpacingType, StyledBaseTheme } from '../Theme'

type PropsType = {
    theme: BaseTheme
    [name: string]: any
}

export type ResolverFunc = (value?: any, props?: PropsType) => any
export const applySingle = (func: ResolverFunc, propName: string, discardNull: boolean = true): any => (props: any) => {
    const { [propName]: propObject } = props
    if (discardNull && !propObject) {
        return null
    }
    return func(propObject, props)
}

type PropsMediaQueryValues<T> = {
    [name in MediaQueryBreakPoint]?: T
}

export type MediaQueryAwareType<T> = T | PropsMediaQueryValues<T>

export type MediaQueryAwareArrayType<T> = MediaQueryAwareType<T | Array<T>>

const applyMediaQueryValues = <T>(
    propValue: MediaQueryAwareArrayType<T> | MediaQueryAwareType<T>,
    func: ResolverFunc, props: PropsType,
) => {
    if (Array.isArray(propValue)) {
        return func(propValue, props)
    }
    return func(propValue, props)
}

export const applyMediaQuery = <T>(func: ResolverFunc, propName: string, discardNull: boolean = true): any => (props: PropsType) => {
    const { [propName]: propObject } = props
    if (discardNull && !propObject) {
        return null
    }
    if (typeof propObject === 'object' && !Array.isArray(propObject)) {
        const mediaPropObject = propObject as PropsMediaQueryValues<T>
        return Object.entries(mediaPropObject).reduce<CssFunction | undefined>((res, [breakPoint, value]): any => css`
            ${res}
            ${MediaQuery[breakPoint as BreakPoint]`
                ${applyMediaQueryValues(value, func, props)}
            `}
        `, undefined)
    }
    return applyMediaQueryValues(propObject, func, props)
}

const resolveSpaceSingle = (space: SpacingType) => space && css`
    padding: ${(p: StyledBaseTheme) => p.theme.spacing(space)};
`
export const resolveSpace = applyMediaQuery(resolveSpaceSingle, 'space')

export const resolveBackgroundColor = (color: Color | string | undefined) => color && css`
    background-color: ${(p: StyledBaseTheme) => p.theme.color(color)};
`

const resolveSingleTextStyle: ResolverFunc = (textSize: number | string, props) => {
    if (typeof textSize === 'string') {
        return `font-size: ${textSize};`
    }
    if (typeof textSize === 'number') {
        const textDesc = props?.theme.textSizes[textSize]
        if (textDesc) {
            return css`
                font-size: ${textDesc.fontSize}${textDesc.unit};
                line-height: ${textDesc.lineHeight}${textDesc.unit};
                font-weight: ${textDesc.fontWeight};
                ${textDesc.textTransform && `text-transform: ${textDesc.textTransform}`}
            `
        }
    }
    return `font-size: ${textSize}px;`
}

export const resolveTextStyle: ResolverFunc = applyMediaQuery(resolveSingleTextStyle, 'textStyle')
