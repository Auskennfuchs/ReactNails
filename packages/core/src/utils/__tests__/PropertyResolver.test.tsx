import { MediaQueryBreakPoint, SpacingType } from '../..'
import { applyMediaQuery } from '../PropertyResolver'

describe('PropertyResolver tests', () => {
    const resolver = (space: SpacingType | SpacingType[]): string => {
        if (Array.isArray(space)) {
            return `padding: ${space.join(' ')}`
        }
        return `padding: ${space}`
    }

    test('should apply single value', () => {
        const res = applyMediaQuery(resolver, 'space')({
            space: SpacingType.m,
        })
        expect(res).toBe('padding: m')
    })

    test('should apply array values', () => {
        const res = applyMediaQuery(resolver, 'space')({
            space: [SpacingType.s, SpacingType.m],
        })
        expect(res).toBe('padding: s m')
    })

    test('should apply media value', () => {
        const res = applyMediaQuery(resolver, 'space')({
            space: {
                [MediaQueryBreakPoint.large]: SpacingType.l,
            },
        })
        expect(res.join('')).toMatchSnapshot()
    })

    test('should apply media array value', () => {
        const res = applyMediaQuery(resolver, 'space')({
            space: {
                [MediaQueryBreakPoint.IE11]: [SpacingType.l, SpacingType.xl],
            },
        })
        expect(res.join('')).toMatchSnapshot()
    })

    test('should apply multiple media values', () => {
        const res = applyMediaQuery(resolver, 'space')({
            space: {
                [MediaQueryBreakPoint.IE11]: [SpacingType.l, SpacingType.xl],
                [MediaQueryBreakPoint.large]: SpacingType.m,
            },
        })
        expect(res.join('')).toMatchSnapshot()
    })
})
