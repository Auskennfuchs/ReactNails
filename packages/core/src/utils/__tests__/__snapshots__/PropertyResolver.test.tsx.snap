// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyResolver tests should apply media array value 1`] = `
"
            
            
@media screen and (-ms-high-contrast: active), (-ms-high-contrast: none) {
    
                padding: l xl
            
}

        "
`;

exports[`PropertyResolver tests should apply media value 1`] = `
"
            
            
@media screen and (min-width: 992px) and (max-width: 1199.98px) {
    
                padding: l
            
}

        "
`;

exports[`PropertyResolver tests should apply multiple media values 1`] = `
"
            
            
            
@media screen and (-ms-high-contrast: active), (-ms-high-contrast: none) {
    
                padding: l xl
            
}

        
            
@media screen and (min-width: 992px) and (max-width: 1199.98px) {
    
                padding: m
            
}

        "
`;
